lib/phase/tree.cpp

```diff
@@ -73,6 +73,11 @@ Result Tree<T>::dispatch(const RecvEndTask &t, TaskQueue &tq, int node_id)
                        Result::DONE_FORWARD));
 }
 
+template <typename T>
+void Tree<T>::mark_init_reached() {
+  reached_nodes[0] = true;
```

Comment:
You removed `reached_nodes[0] = true` from one place and put the same code in multiple other places. This is generally not a good pattern.